Rework Accessing Virtual Machines section (#599)
kubevirt/user-guide: docs/virtual_machines/accessing_virtual_machines.md

## Graphical and Serial Console Access

Once a virtual machine is started you are able to connect to the consoles it
exposes. Usually there are two types of consoles:

- Serial Console
- Graphical Console (VNC)

> Note: You need to have `virtctl`
> [installed](../operations/virtctl_client_tool.md) to gain
> access to the VirtualMachineInstance.

### Accessing the Serial Console

The serial console of a virtual machine can be accessed by using the `console`
command:

```shell
virtctl console testvm
```

### Accessing the Graphical Console (VNC)

To access the graphical console of a virtual machine the VNC protocol is
typically used. This requires `remote-viewer` to be installed. Once the
tool is installed, you can access the graphical console using:

```shell
virtctl vnc testvm
```

If you only want to open a vnc-proxy without executing the `remote-viewer`
command, it can be accomplished with:

```shell
virtctl vnc --proxy-only testvm
```

This would print the port number on your machine where you can manually connect
using any VNC viewer.

### Debugging console access

> Should the connection fail, you can use the `-v` flag to get more verbose

Author: Changed

output from both `virtctl` and the `remote-viewer` tool to troubleshoot the
problem.

```shell
virtctl vnc testvm -v 4
```

> **Note:** If you are using virtctl via SSH on a remote machine, you need to
> forward the X session to your machine. Look up the -X and -Y flags of `ssh` if
> you are not familiar with that. As an alternative you can proxy the API server
> port with SSH to your machine (either direct or in combination
> with `kubectl proxy`).

## SSH Access

A common operational pattern used when managing virtual machines is to inject
SSH public keys into the virtual machines at boot. This allows automation tools
(like Ansible) to provision the virtual machine. It also gives operators a way
of gaining secure and passwordless access to a virtual machine.

KubeVirt provides multiple ways to inject SSH public keys into a virtual
machine. In general, these methods fall into two categories.
[Static key injection](./accessing_virtual_machines.md#static-ssh-key-injection-via-cloud-init)

> , which places keys on the virtual machine the first time it is booted, and

Author: Changed

[dynamic injection](./accessing_virtual_machines.md#dynamic-ssh-key-injection-via-qemu-user-agent)

> , which allows keys to be dynamically updated both at boot and during runtime.

Author: Changed

> KubeVirt provides multiple ways to inject SSH public keys into a virtual
> machine. In general, these methods fall into two categories.
> [Static key injection](./accessing_virtual_machines.md#static-ssh-key-injection-via-cloud-init),
> which places keys on the virtual machine the first time it is booted, and
> [dynamic injection](./accessing_virtual_machines.md#dynamic-ssh-key-injection-via-qemu-user-agent),
> which allows keys to be dynamically updated both at boot and during runtime.

Reviewer: ```suggestion
KubeVirt provides multiple ways to inject SSH public keys into a virtual
machine. 

In general, these methods fall into two categories:
- [Static key injection](./accessing_virtual_machines.md#static-ssh-key-injection-via-cloud-init),
which places keys on the virtual machine the first time it is booted
- [dynamic injection](./accessing_virtual_machines.md#dynamic-ssh-key-injection-via-qemu-user-agent),
which allows keys to be dynamically updated both at boot and during runtime.
```

> Once a SSH public key is injected into the virtual machine, access to it can be
> gained with the help of `virtctl`.

Reviewer: ```suggestion
Once a SSH public key is injected into the virtual machine, it can be
accessed via `virtctl`.
```

### Static SSH public key injection via cloud-init

> Users creating virtual machines have the ability to provide startup scripts to
> their virtual machines, which allow any number of customization operations to
> take place. Placing SSH public keys into a
> [cloud-init startup script](./startup_scripts.md#cloud-init)
> is one option for injecting them into a virtual machine. However, there are
> other, more flexible options available.

Reviewer: ```suggestion
Users creating virtual machines can provide startup scripts to
their virtual machines, allowing multiple customization operations.

One option for injecting public SSH keys into a VM is via [cloud-init startup script](./startup_scripts.md#cloud-init). However, there are more flexible options available.
```

The virtual machine's access credential API allows statically injecting SSH
public keys at startup time independently of the cloud-init user data by
placing the SSH public key into a Kubernetes `Secret`. This allows to keep the
application data in the cloud-init user data separated from the credentials
used to access the virtual machine.

A Kubernetes `Secret` can be created from an SSH public key like this:

```shell
# Place SSH public key into a Secret
kubectl create secret generic my-pub-key --from-file=key1=id_rsa.pub
```

The `Secret` containing the public key is then assigned to a virtual machine

> using the access credentials API with the `configDrive` propagation method.

Reviewer: @qinqon I remember discussing w/ you some time ago; does it really need to be `configDrive` ? I have the slight recollection you told me this would also work w/ `NoCloud` . 

Did I misunderstand you / am I wrong ?

Author: It's only implemented for `configDrive` at the moment.

See variable `keys` that is injected into the config drive metadata:
https://github.com/kubevirt/kubevirt/blob/4de7e9773109226c1ecc9e4a0df9ed2f1f6b0cd5/pkg/cloud-init/cloud-init.go#L137

Reviewer: Thanks. Looking at the code and at the [current README](https://kubevirt.io/user-guide/virtual_machines/startup_scripts/#injecting-ssh-keys-with-cloud-inits-cloud-config) I think I misunderstood: we can sneak the public keys via secret via the `userData` which is defined in a k8s secret. 

Thanks for the code link - the truth (as always ...) is in the code.

Reviewer 2: That's weird this is working for us on capk, https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/185/files#diff-bab123ad42a3d34ff93b21e3c9695fb2d81b2ea4bcd87eaea42b987565153bdaR152

KubeVirt injects the SSH public key into the virtual machine by using the
generated cloud-init metadata instead of the user data. This keeps the
application respectively user data separated from credentials.

> Note: The cloud-init `userData` is not touched.

```shell
# Create a VM referencing the Secret using propagation method configDrive
kubectl create -f - <<EOF
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: testvm
spec:

>   running: false

Author: Good idea!

          propagationMethod:
            qemuGuestAgent:
              users:
              - fedora
      volumes:
      - containerDisk:
          image: quay.io/containerdisks/fedora:latest
        name: containerdisk
      - cloudInitConfigDrive:
          userData: |-
            #cloud-config
            password: fedora
            chpasswd: { expire: False }
            # Disable SELinux for now, so qemu-guest-agent can write the authorized_keys file
            bootcmd:

>               - setenforce 0

Reviewer 3: Can we avoid using the SELinux permissive mode by assigning the appropriate SELinux context?
Something similar to:
 `- ["sudo", "chcon", "--type", "virt_qemu_ga_exec_t", "/home/$USER/.ssh/authorized_keys"]`

Author: The issue lies deeper, out of the context for this example.

`audit2why` says:

```
type=AVC msg=audit(1665750224.275:505): avc:  denied  { dac_read_search } for  pid=674 comm="qemu-ga" capability=2  scontext=system_u:system_r:virt_qemu_ga_t:s0 tcontext=system_u:system_r:virt_qemu_ga_t:s0 tclass=capability permissive=0
type=AVC msg=audit(1665750224.275:506): avc:  denied  { dac_override } for  pid=674 comm="qemu-ga" capability=1  scontext=system_u:system_r:virt_qemu_ga_t:s0 tcontext=system_u:system_r:virt_qemu_ga_t:s0 tclass=capability permissive=0
```

I would rather keep disabling SELinux for now. (although I really don't like to)

Reviewer 3: I agree.

        name: cloudinitdisk
EOF
```

### Accessing SSH with virtctl

Every `VirtualMachineInstance` provides a `virtualmachineinstances/portforward`
subresource, that can be used to create a websocket backed network tunnel to
a port inside the instance similar to Kubernetes pods.

One use-case for this subresource is to forward SSH traffic into
the `VirtualMachineInstance` either from the CLI or a web-UI.

To connect to a `VirtualMachineInstance` from your local machine, `virtctl`
provides a lightweight SSH client with the `ssh` command, that uses port
forwarding. Refer to the command's help for more details.

```shell

> virtctl ssh

Reviewer 3: ```suggestion
virtctl ssh <virtual-machine-instance-name>
```
Or something similar that makes clear that you have to provide the VMI name.

Author: I deliberately left it out so the mentioned help would get displayed.

```

To transfer files from or to a `VirtualMachineInstance` `virtctl` also
provides a lightweight SCP client with the `scp` command. Its usage is
similar to the `ssh` command. Refer to the command's help for more details.

```shell

> virtctl scp

Reviewer 3: ```suggestion
virtctl scp  <virtual-machine-instance-name>
```

Author: I deliberately left it out so the mentioned help would get displayed.

   port to your local stdout/stdin. This allows the command to be used in
   combination with the OpenSSH client's `ProxyCommand` option.

```shell
ssh -o 'ProxyCommand=virtctl port-forward --stdio=true vmi/testvm.mynamespace 22' fedora@testvm.mynamespace
```

To provide easier access to arbitrary virtual machines you can add the following
lines to your SSH `config`:

```
Host vmi/*
   ProxyCommand virtctl port-forward --stdio=true %h %p
Host vm/*
   ProxyCommand virtctl port-forward --stdio=true %h %p
```

This allows you to simply call `ssh user@vmi/testvmi.mynamespace` and your
SSH config and virtctl will do the rest. Using this method it becomes easy
to set up different identities for different namespaces inside your SSH
`config`.

Note that all port forwarding traffic will be sent over the Kubernetes
control plane. A high amount of connections and traffic can increase
pressure on the API server. If you regularly need a high amount of connections
and traffic consider using a dedicated Kubernetes `Service` instead.

#### Example

1. Create virtual machine and inject SSH public key as explained above

2. Start virtual machine

```shell

> virtctl start testvm

Reviewer 3: If we agree to set `running: true` this is not required.

Author: Yes, I removed it.

          propagationMethod:
            qemuGuestAgent:
              users:
              - fedora
      volumes:
      - containerDisk:
          image: quay.io/containerdisks/fedora:latest
        name: containerdisk
      - cloudInitConfigDrive:
          userData: |-
            #cloud-config
            password: fedora
            chpasswd: { expire: False }
            # Disable SELinux for now, so qemu-guest-agent can write the authorized_keys file
            # The selinux-policy is too restrictive currently, see open bugs:
            #   - https://bugzilla.redhat.com/show_bug.cgi?id=1917024
            #   - https://bugzilla.redhat.com/show_bug.cgi?id=2028762
            #   - https://bugzilla.redhat.com/show_bug.cgi?id=2057310
            bootcmd:
              - setenforce 0
        name: cloudinitdisk
EOF
```

> ### Accessing SSH with virtctl

Author: I changed it to `Accessing a VMI using virtctl` to be a bit less specific. The abstract below talks about "every" VMI.

> Every `VirtualMachineInstance` provides a `virtualmachineinstances/portforward`
> subresource, that can be used to create a websocket backed network tunnel to
> a port inside the instance similar to Kubernetes pods.

Reviewer: I personally prefer if the sentence's subject is the user, rather than the VMI. Something like:
```suggestion
The user can create a websocket backed network tunnel to
a port inside the instance by using the VirtualMachineInstance virtualmachineinstances/portforward
subresource.
```

similar to the `ssh` command. Refer to the command's help for more details.

```shell
virtctl scp
```

#### Using virtctl as proxy

If you prefer to use your local OpenSSH client, there are two ways of doing
that in combination with virtctl.

> Note: Most of this applies to the `virtctl scp` command too.

1. The `virtctl ssh` command has a `--local-ssh` option. With this option
   `virtctl` wraps the local OpenSSH client transparently to the user. The
   executed SSH command can be viewed by increasing the verbosity (`-v 3`).

```shell
virtctl ssh --local-ssh -v 3 testvm
```

2. The `virtctl port-forward` command provides an option to tunnel a single
   port to your local stdout/stdin. This allows the command to be used in
   combination with the OpenSSH client's `ProxyCommand` option.

```shell
ssh -o 'ProxyCommand=virtctl port-forward --stdio=true vmi/testvm.mynamespace 22' fedora@testvm.mynamespace
```

To provide easier access to arbitrary virtual machines you can add the following
lines to your SSH `config`:

```
Host vmi/*
   ProxyCommand virtctl port-forward --stdio=true %h %p
Host vm/*
   ProxyCommand virtctl port-forward --stdio=true %h %p
```

> This allows you to simply call `ssh user@vmi/testvmi.mynamespace` and your

Reviewer 4: Maybe in a follow-up we can highlight how to use a ssh config file in combination with ansible.

Something like:

```
[ssh_connection]
ssh_args = -F ~/.ssh/myconfig
```

Then people can write their inventory files like always and provision VMs over the mgmt-network :)

Author: Good call, I added a small paragraph describing this.
